### DISCUSSION TITLE
wait time of downloading is increased

### DISCUSSION THREAD

**Discussion by highestbest** [aclap-dev/video-downloadhelper#829] [Help and Q&A]
### Your question

for example, i want to download 2 file. (01.mp4, 02.mp4)
at first, i try to download 01.mp4.
Next, i set the path for D://. then, i start to download at D://01.mp4 path.
i try to download 02.mp4 before finishing download 01.mp4.
i have to wait to setting for path until i finish downloading 01.mp4.

this case is 2 download.
but if download file is lot of number, it is not easy.
wait time is increased.
thus, i ask you to enable to check or fix it.

### Addon & CoApp version and details

version: 9.0.2.3
target: mozilla
channel: stable
lang: ko
coapp: {"found":true,"path":"C:\\Program Files\\DownloadHelper CoApp\\vdhcoapp.exe","version":"2.0.19","new_version":false}
license: {"unset":true}
platform: x86-64 win
UA: Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:109.0) Gecko/20100101 Firefox/115.0
{
    "used_sidebar_button": true,
    "used_popup_button": true,
    "download_directory": "D:\\Movie_J",
    "last_download_directory": {
        "__serializer_tag": "option_some",
        "__serializer_value": {
            "__serializer_tag": "primitive",
            "__serializer_value": "F:\\"
        }
    },
    "concurrent_downloads_max": 1,
    "show_success_notification_for_icognito": false,
    "never_show_no_incognito_msg_again": true,
    "theme": "system",
    "view_options": {
        "all_tabs": false,
        "low_quality": false,
        "sort_by_status": true,
        "sort_reverse": false,
        "show_button_clean": false,
        "show_button_clean_all": true,
        "show_button_convert_local": false,
        "hide_downloaded": true
    },
    "use_sidebar": false,
    "open_count_store": 12,
    "successfull_dl": 24
}

**Comment by GrampaWildWilly** [comment]
![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/6a7d5d4c-de4e-4121-8e65-655d94f45ba8)
![#02](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/1d22b38e-1036-407e-ac04-66b02f0fd915)
![#03](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/7c80c819-aa6b-42cc-a102-9b613a4440a7)

**Reply by GrampaWildWilly** [reply]
I do not work for VDH.  I'm just another user like you.  I can't answer your question.

**Reply by GrampaWildWilly** [reply]
One reason I can't answer your question is I don't understand it.  I offered what I thought might be your solution but apparently it is not.  So now I realize I don't understand what your problem is.  Please describe your problem using screenshots.  Show the steps of what you try to do & explain how this isn't working the way you want.  Do not try to explain your problem in words alone.  Add screenshots to illustrate what you are trying to do.

**Reply by highestbest** [reply]
my previous reply was deleted by my mistake. I'm sorry.

To illustrate the phenomenon, i add a screenshot as follows.
As i mentioned, my computer is very old, i want to get it one by one. Thus i do setting of 1 max concurrent download.

![test1](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/61a58512-10f9-4e4a-8df9-17ab44b33048)
capture1. first download

![test2](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/63012b25-aca2-40a2-ad3c-6b95bf0ef202)
capture2. first download path

![test3](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/e5933c49-fca5-4f6c-8175-62e0962119a5)
capture3. first download name

![test4](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/0aa34b45-e831-4e94-8ecb-e15be6d5e3d5)
capture4. second download

![test5](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/bed8d65e-003b-458d-a60f-608fab8e0948)
capture5. second download path

![test6](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/bcfb5d00-075f-4f0a-85e1-d01fd4a4f5d6)
capture6. second download path is not working before first download is finished

![test7](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/87eb2b8e-e7e5-47f4-9c0b-09e845da49d4)
capture7. second download name is working after first download is finished

download didn't start until one is finished, they only start when one is finished.
it is working old version. but, new version is not working.
I want to ask VDH why it is not working. i am sorry if there was any misunderstanding.

**Reply by GrampaWildWilly** [reply]
I think I understand now.  While the first download is in progress, VDH doesn't even let you queue up the second request.  You would like to queue up several requests but let VDH download them one at a time.  Do I have this correct now?

There is another discussion in this forum (#139) in which a user is queuing up several download requests but his complaint is that VDH doesn't select the next download in the order in which he queued it.  But he is getting multiple downloads in his queue.  So he is not seeing the problem you are having.  But that was with a beta version of VDH 8.  You are using VDH 9.  That other thread suggests that @paulrouget Paul revamped the queuing mechanism in VDH 9.  I would suggest trying the latest beta VDH for Firefox.  It's not clear from that other discussion, which does not seem to be resolved yet, whether the problem has actually been addressed in VDH 9.  Still, the beta is worth a try.  Click the link at the top of this page labelled Wiki.  That is the closest thing we have to a user manual.  You will see chapters listed down the right side of the page.  Click the one that discusses beta versions.  You will find the VDH beta there.  I am running 9.0.2.3a1 beta, which is the latest at the moment.  Maybe that will cure your problem.  If it does not, you have reported a valuable problem that Paul would need to address.

**Reply by highestbest** [reply]
yes. it is correct to understand.
i remember that the order change phenomenon was recognized and fixed at old version (probably 8.x), but there was no way to check it because it was not working second request in this version (9.0.2.3). i'll try to test the beta version.

**Reply by highestbest** [reply]
i use 9.0.2.3a1 beta version. but it is same situation.

![test1](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/c2111d59-56f3-42cb-9a81-e41d90fd8ca5)
figure 1. 9.0.2.3a1 version

![test2](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/46d5997d-cbeb-490c-bcaa-657670cdb525)
figure 2. next download path is not working before first download is finished

let me know if i have misunderstanding
thank you for your help in advance

**Reply by GrampaWildWilly** [reply]
That should work.  It appears that you are hitting a bug in the product.  Like I said before, I do not work for VDH.  I'm just another user.  So I can't fix anything.  I mentioned @paulrouget Paul upthread here.  He is the developer.  He is the only one who can correct this.  For now, I would suggest that you just go ahead & increase your Max concurrent downloads setting, like I described in my first answer here.  It may be an old computer but if you want to be able to get multiple downloads at least started, this is probably the only thing you can do until Paul fixes the problem.  Either that or just wait until your first download completes before you start the next one.  I know that's not what you want, but what you want doesn't work right now.  So quit trying to fight it.  It's not working.  You have to wait for Paul to fix the program.

**Reply by highestbest** [reply]
yes. i don't try to fight.
i will wait for Paul to fix it, not you.
thank you.

**Reply by paulrouget** [reply]
I have pushed 9.0.2.12.

Does this improve the situation?

**Reply by highestbest** [reply]
i am so glad to meet you, Paul.
unfortunately, this situation is not improved.

![test1](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/f7090e35-7905-4103-9da1-b8f67a1a771a)
figure 1. 9.0.2.12 version

![test2](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/7be396b4-6438-4b4f-b341-b7ca2026b7bc)
figure 2. next download path is not working before first download is finished

![test3](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/a1d3858b-e5a4-48d3-a232-1fb04b475096)
figure 3. first download is finished

![test4](https://github.com/aclap-dev/video-downloadhelper/assets/171928439/dac685bd-1a62-4f92-a7ea-a9650c63fa6e)
figure 4. next download name is working after first download is finished

i know another problems, but I think that you feel confusion if i speak what this is.
thus, i only focus this problem.
after it is solved, i'll speak next problem.

**Comment by GrampaWildWilly** [comment]
Using Windows 11 Pro 64-bit version 23H2 build 22631.3737, Firefox 127.0.1 64-bit, licensed VDH 9.0.2.12a1 beta, CoApp 2.0.19, I have replicated this problem.  Here's how.

Go into VDH More settings & set Max concurrent downloads to 1.  With that set, you can launch one download but a second download request is not queued.  It is blocked.  The progress on the second download remains stuck with the spinner on what will eventually become the Stop button.  Once the first download completes, the file picker for the second download appears.  So queuing is not possible because only 1 download is allowed to even be launched at a time.  This is not correct.  You should be able to launch the first download & while it is in progress, you should be able to do another Download As & go through the file picker dialog.  That request should be queued, and then launch once the first download completes.  Ditto for any further downloads you might request while the first download is still in progress.  Then one by one, the queued requests should be serviced.  Plus, to pick up the issue from another discussion, those queued requests should be serviced in the same order you issued the requests.

The only other value for Max concurrent downloads I tested was 16.  With that value, you can make multiple requests.  I managed to get 4 downloads running concurrently.  I suspect I might have hit a problem if I had managed to get 16 downloads running concurrently & tried to queue up a 17th one.  You should test that boundary condition, Paul.  My suspicion is that when you reach whatever your Max concurrent download limit is, you will be blocked instead of being able to just keep queuing up more downloads.  If your setting is 1, you're blocked after just one.  I haven't tested it but if it's 2, your 3rd attempted download will probably be blocked.  And so on.  You've got the code, Paul.  Read it & see if that's what it's doing.

Our original poster here wants his setting to have the value 1.  His hardware is not up to doing multiple concurrent downloads.  But he should at least be able to queue up additional downloads while his one download might be in progress.  This is not possible with the current VDH.

**Reply by GrampaWildWilly** [reply]
VDH recognizes a download as having started even if you've only started the file selection.  The file selection dialog is treated within Windows as a separate task.  You can `Alt+Tab` to it & `Alt+Tab` away from it.  So you can, for example, start a `Download As` for one video, get the file selection dialog started, then `Alt+Tab` to a different video.  The file selection dialog will still be open, uncompleted.  That download will not actually be in progress yet.  It will be pending your navigating to a directory & either accepting or changing the file name.  But VDH will count it as a started download.  You can then start a second `Download As` & leave it sitting on its file selection dialog while you go to yet a third video & partially start that process.  VDH will show a 3 within a blue spot on its icon on the tool bar of the Firefox window.  There will not be 3 downloads really being transferred.  There will just be 3 file selection processes pending.  That will be 3 open file selection windows waiting for you to complete them.  As you complete each one, a download will actually begin.

But all of that is possible only if you have `Max concurrent downloads` set at a high enough value.  If you have that setting at 1, then you can get ONLY 1 file selection dialog pending.  Your download will not actually be in progress but VDH will have added 1 to its count of active downloads.  That means that even though you can `Alt+Tab` to another video & start a `Download As`, that second `Download As` will stay stuck with the spinner in the VDH menu.  Its file selection dialog will NOT launch until VDH completes the first file selection dialog & actually starts that download.  That download then has to complete before the second file selection dialog will open.

So the problem appears to be the internal definition inside VDH of what is a "download in progress."  Just beginning to define a download counts today as a started download.  This is not correct.  VDH should allow an unlimited number of attempts to start a download.  It should apply the `Max concurrent download` setting ONLY to downloads that are truly in the process of transferring data over your Internet connection.  Inflight download definitions prior to launch of actual data transfer should NOT count as started downloads.

Adding 1 to the count inside the blue spot just by initiating a `Download As` has always been the way VDH worked, that is to say, in past releases when the terminology was different.  But somehow, VDH always applied the `Max concurrent downloads` setting only to the downloads that really had data coming across the Internet connection.  You used to be able to queue up as many additional downloads as you wanted.  `Max concurrent downloads` would apply only to the number of Internet connections VDH was managing, not the number of defined, queued up, waiting downloads you might have requested.  My interpretation of the problem is that VDH is now relying on the wrong internal counter to block the addition of download requests to the queue.  There should be only 1 counter & it should apply only to processes that really are transferring data over the Internet connection.  There should be NO counter for  `Download As` functions that are in the process of defining additional downloads that are destined to be queued rather than simply started.
